gateway/ln-gateway/src/rpc/lnrpc_client.rs

```diff
@@ -0,0 +1,126 @@
+use std::{fmt::Debug, net::SocketAddr, sync::Arc};
+
+use async_trait::async_trait;
+use fedimint_api::{dyn_newtype_define, task::TaskGroup};
+use fedimint_server::modules::ln::contracts::Preimage;
+use lightning_invoice::Invoice;
+use secp256k1::PublicKey;
+use tokio::sync::{mpsc, Mutex};
+use tonic::transport::Channel;
+
+use super::HtlcInterceptPayload;
+use crate::{gatewaylnrpc::gateway_lightning_client::GatewayLightningClient, Result};
+
+#[derive(Debug, Clone)]
+
+pub struct InvoiceInfo {
+    pub invoice: Invoice,
+    pub max_delay: u64,
+    pub max_fee_percent: f64,
+}
+
+/**
+ * Convenience wrapper around `GatewayLightningClient` protocol spec
+ * This provides ease in constructing rpc requests and parsing responses.
+ */
+#[async_trait]
+pub trait ILnRpcClient: Debug + Send + Sync {
+    // Get the public key of the lightning node
+    async fn get_pubkey(&self) -> Result<PublicKey>;
+
+    // Attempt to pay an invoice using the lightning node
+    async fn pay_invoice(&self, invoices: Vec<InvoiceInfo>) -> Result<Vec<Result<Preimage>>>;
+
+    // Subscribe to intercept htlcs that belong to a specific mint identified by `short_channel_id`
+    async fn subscribe_intercept_htlcs(
+        &self,
+        short_channel_id: u64,
+    ) -> Result<mpsc::Receiver<HtlcInterceptPayload>>;
+}
+
+dyn_newtype_define!(
+    /// Arc reference to a gateway lightning rpc client
+    #[derive(Clone)]
+    pub LnRpcClient(Arc<ILnRpcClient>)
+);
+
+impl LnRpcClient {
+    pub fn new(client: Arc<dyn ILnRpcClient + Send + Sync>) -> Self {
+        LnRpcClient(client)
+    }
+}
+
+/**
+ * An `ILnRpcClient` that wraps around `GatewayLightningClient` for convenience,
+ * and makes real RPC requests over the wire to a remote lightning node.
+ * The lightnign node is exposed via a corresponding `GatewayLightningServer`.
+ */
+#[derive(Debug)]
+pub struct NetworkLnRpcClient {
+    client: Mutex<GatewayLightningClient<Channel>>,
+    task_group: TaskGroup,
+}
```

Comment:
I'm still not sure if we really want or need that task group and the mutex-ed client, but my confusion is less with this struct and more with "wouldn't it be nicer to just use the client directly instead of through the task and piping things through channels"

Reply:
Right, when writing the MVP, interfacing the actor and the client directly seemed "nasty and inconvenient". Seemed to me like I could have this client as an intermediator between raw grpc and the rust types - but I agree the cost of having this wrapper is quite high, especially if we have to map over the stream and then send received messages over anotehr channel to gateway actors.

I think I've learned enough to figure out how to use the raw grpc client directly. and skip this wrapper client all together

Reply:
Okay might still need a wrapping struct here to hide the details of Transport choice from gatewayd.
The protogen auto defined client struct is generic over transport mechanisms : `GatewayLightningClient<T>`
 - In production, I want a client with [tonic::transport::Channel](https://docs.rs/tonic/0.1.0-alpha.2/tonic/transport/struct.Channel.html) for making network grpc calls to remote [lightning extension service](https://github.com/fedimint/fedimint/pull/1295). This encourages me to write `NetworkLnRpcClient` which composes a `GatewayLightningClient<tonic::transport::Channel>` for sending and receiving requests. Latest learnings:
   -  `NetworkLnRpcClient` doesn't need a mutex ref to the composed client
   - `NetworkLnRpcClient` should simply pass out the response it gets from it's composed client, so there's no need for adapting the tonic streams received, out into some channel messaging system for gatewayd processing
 - In test scenarios, I don't need a transport. Rather, I want to mock out the whole mechanism and present the same API interface as `GatewayLightningClient<T>`
 
An additional requirement is that I need to be able to construct the grpc client on the fly at gatewayd runtime. This allows configuring of the dependency between gatewayd and remote lightning after boot of gatewayd, for example, through some ui.

Working on updates